src/front/wgsl/lexer.rs

```diff
@@ -1,267 +1,237 @@
+            }
+        }
+        match kind {
+            "" => match i64::from_str_radix(input, radix) {
+                Ok(num) => Ok(Number::Int(num)),
+                Err(e) => Err(map_err(e)),
             },
-
-            NumberLexerState {
-                digit_state: NLDigitState::Exponent,
-                ..
-            } => match c {
-                '0'..='9' => {
-                    state.digit_state = NLDigitState::DigitAfterExponent;
-                }
-                '-' | '+' => {
-                    state.digit_state = NLDigitState::SignAfterExponent;
-                }
-                _ => return false,
+            "i" => match i32::from_str_radix(input, radix) {
+                Ok(num) => Ok(Number::I32(num)),
+                Err(e) => Err(map_err(e)),
             },
-
-            NumberLexerState {
-                digit_state: NLDigitState::SignAfterExponent | NLDigitState::DigitAfterExponent,
-                ..
-            } => match c {
-                '0'..='9' => {
-                    state.digit_state = NLDigitState::DigitAfterExponent;
-                }
-                _ => return false,
+            "u" if is_negative => Err(NumberError::NotRepresentable),
+            "u" => match u32::from_str_radix(input, radix) {
+                Ok(num) => Ok(Number::U32(num)),
+                Err(e) => Err(map_err(e)),
             },
+            _ => unreachable!(),
         }
+    }
 
-        // No match branch has rejected this yet, so we are still in a number literal
-        true
-    };
-
-    let pos = working_substr
-        .find(|c| !what(c))
-        .unwrap_or(working_substr.len());
-    let (value, rest) = input.split_at(pos + minus_offset + hex_offset);
-
-    // NOTE: This code can use string slicing,
-    //       because number literals are exclusively ASCII.
-    //       This means all relevant characters fit into one byte
-    //       and using string slicing (which slices UTF-8 bytes) works for us.
+    // check for hex first
+    if lazy_regex::regex_is_match!(r"^-?0[xX].*"sx, input) {
+        if let Some((_, sign, significand, exponent, kind, rest)) = lazy_regex::regex_captures!(r"
+                ^(-)?    0[xX]    ([0-9a-fA-F]+\.[0-9a-fA-F]* | [0-9a-fA-F]*\.[0-9a-fA-F]+)    (?:([pP][+-]?[0-9]+)    ([fh]?))?    (.*)
+            "sx, input)
+        {
+            // | / 0[xX][0-9a-fA-F]+    \.[0-9a-fA-F]*      ([pP][+-]?[0-9]+        [fh]?)? /
+            // | / 0[xX][0-9a-fA-F]*    \.[0-9a-fA-F]+      ([pP][+-]?[0-9]+        [fh]?)? /
+
+            // 0[xX] and [pP] is required by hexf
+            let hexf_input = &format!(
+                "{}0x{}{}",
+                sign,
+                significand,
+                if exponent.is_empty() { "p0" } else { exponent }
+            );
+
+            (parse_hex_float(hexf_input, kind), rest)
+        } else if let Some((_, sign, significand, exponent, kind, rest)) = lazy_regex::regex_captures!(r"
```

Comment:
Is using "lazy" semantics here really warranted? This is effectively an implicit global state with initialization at the first run. Since we know all of the regex patterns in advance, alternatively we could just build them all at the creation of WGSL parser and store with it. This would make our first shader compilation predictable, and as a small bonus avoid an extra dependency on `lazy_regex`.

Reply:
I've updated the code in this direction. The fact that we now need to pass the `NumberRegexes` struct around isn't as nice as `lazy_regex`'s API but I agree it's not really warranted to add another dep. (Also `lazy_regex` doesn't support `RegexSet`s)